test/test_invalid_headers.py

```diff
@@ -92,18 +95,56 @@ class TestFilter(object):
     HTTP/2 and so may never hit the function, but it's worth validating that it
     behaves as expected anyway.
     """
-    @given(HEADERS_STRATEGY)
-    def test_range_of_acceptable_outputs(self, headers):
+    true_false_combinations = [
+        (True, True),
+        (True, False),
+        (False, True),
+        (False, False),
+    ]
+
+    @pytest.mark.parametrize('is_client,is_trailer', true_false_combinations)
+    @given(headers=HEADERS_STRATEGY)
+    def test_range_of_acceptable_outputs(self, headers, is_client, is_trailer):
         """
         validate_headers either returns the data unchanged or throws a
         ProtocolError.
         """
         try:
-            assert headers == h2.utilities.validate_headers(headers)
+            conn_state = h2.utilities.HeaderValidationFlags(
+                is_client=is_client,
+                is_trailer=is_trailer,
+            )
```

Comment:
We're at the height of pedanticness now, but seeing as you've been so willing to tinker...can you hoist the construction of the `HeaderValidationFlags` both here and in `test_invalid_pseudo_headers` out of their containing exception scopes? It makes it a bit less clear what statement we're trying to protect. =) Small `try` blocks are good.